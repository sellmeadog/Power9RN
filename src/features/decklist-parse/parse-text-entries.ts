import { P9DecklistEntryType } from '../../core/data-user';
import { P9CreateDecklistEntryInfo } from './parse-document-types';

const REGEX_EMPTY_LINE = /^[\r|\n]*?$/im;
const REGEX_SIDEBOARD_LINE = /^(?:.*?Sideboard.*?)?$/im;
const REGEX_DECKLIST_ENTRY_LINE = /^((?:SB:\s*?)?\d+)\s*(.*(?=\s\(\w*?\))|.*)(?:\s\((\w*)\)\s?(\w*)?)?$/gim;

export function parseEntryInfo(type: P9DecklistEntryType, textEntries: string): P9CreateDecklistEntryInfo[] {
  const entries: P9CreateDecklistEntryInfo[] = [];
  let match: RegExpExecArray | null;

  while ((match = REGEX_DECKLIST_ENTRY_LINE.exec(textEntries))) {
    const [_, count, cardName, expansionCode, collectorNumber] = match;

    entries.push({
      type,
      count,
      cardName,
      expansionCode,
      collectorNumber,
    });
  }

  return entries;
}

export const parseTextEntries = (documentContent: string) => {
  const [maindeck, sideboard] = documentContent
    .trim()
    .split(REGEX_SIDEBOARD_LINE)
    .filter((s) => !s.match(REGEX_EMPTY_LINE));

  const maindeckEntryInfo = parseEntryInfo('maindeck', maindeck);
  const sideboardEntryInfo = parseEntryInfo('sideboard', sideboard);

  return [...maindeckEntryInfo, ...sideboardEntryInfo];
};
