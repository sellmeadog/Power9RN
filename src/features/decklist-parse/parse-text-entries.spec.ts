import { parseTextEntries } from './parse-text-entries';

describe('parseTextEntries', () => {
  it('should parse standard format text entries', () => {
    // Arrange
    const textEntries = `
4 Gravecrawler
4 geralfs messenger
3 damping sphere
`;

    // Act
    const actual = parseTextEntries(textEntries);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse standard format text entries with empty line divider', () => {
    // Arrange
    const textEntries = `
4 Gravecrawler
4 geralfs messenger

3 damping sphere
`;

    // Act
    const actual = parseTextEntries(textEntries);

    // Assert
    // expect(
    //   actual.includes({
    //     cardName: 'damping sphere',
    //     collectorNumber: undefined,
    //     count: '3',
    //     expansionCode: undefined,
    //     type: 'sideboard',
    //   })
    // ).toBeTruthy();
    expect(actual).toMatchSnapshot();
  });

  it('should parse standard format text entries with Sideboard line divider', () => {
    // Arrange
    const textEntries = `
4 Gravecrawler
4 geralfs messenger
Sideboard:
3 damping sphere
`;

    // Act
    const actual = parseTextEntries(textEntries);

    // Assert
    // expect(
    //   actual.includes({
    //     cardName: 'damping sphere',
    //     collectorNumber: undefined,
    //     count: '3',
    //     expansionCode: undefined,
    //     type: 'sideboard',
    //   })
    // ).toBeTruthy();
    expect(actual).toMatchSnapshot();
  });

  it('should parse standard format text entries with Sideboard line and extra whitespace', () => {
    // Arrange
    const textEntries = `
4 Gravecrawler
4 geralfs messenger


Sideboard:
3 damping sphere
`;

    // Act
    const actual = parseTextEntries(textEntries);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse standard format text entries regardless of bad input', () => {
    // Arrange
    const textEntries = `
4 Gravecrawler
chili cheese dogs
4 geralfs messenger
Sideboard:
3 damping sphere
`;

    // Act
    const actual = parseTextEntries(textEntries);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should gracefully handle bad input', () => {
    // Arrange
    const textEntries = `
this is a useless
Sideboard:
list of card entries
in a deck list
`;

    // Act
    const actual = parseTextEntries(textEntries);

    // Assert
    expect(actual).toMatchSnapshot();
  });
});
