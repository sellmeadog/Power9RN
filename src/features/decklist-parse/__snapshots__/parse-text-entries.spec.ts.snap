// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTextEntries should gracefully handle bad input 1`] = `Array []`;

exports[`parseTextEntries should parse standard format text entries 1`] = `
Array [
  Object {
    "cardName": "Gravecrawler",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "geralfs messenger",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "damping sphere",
    "collectorNumber": undefined,
    "count": "3",
    "expansionCode": undefined,
    "type": "maindeck",
  },
]
`;

exports[`parseTextEntries should parse standard format text entries regardless of bad input 1`] = `
Array [
  Object {
    "cardName": "Gravecrawler",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "geralfs messenger",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "damping sphere",
    "collectorNumber": undefined,
    "count": "3",
    "expansionCode": undefined,
    "type": "sideboard",
  },
]
`;

exports[`parseTextEntries should parse standard format text entries with Sideboard line and extra whitespace 1`] = `
Array [
  Object {
    "cardName": "Gravecrawler",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "geralfs messenger",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "damping sphere",
    "collectorNumber": undefined,
    "count": "3",
    "expansionCode": undefined,
    "type": "sideboard",
  },
]
`;

exports[`parseTextEntries should parse standard format text entries with Sideboard line divider 1`] = `
Array [
  Object {
    "cardName": "Gravecrawler",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "geralfs messenger",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "damping sphere",
    "collectorNumber": undefined,
    "count": "3",
    "expansionCode": undefined,
    "type": "sideboard",
  },
]
`;

exports[`parseTextEntries should parse standard format text entries with empty line divider 1`] = `
Array [
  Object {
    "cardName": "Gravecrawler",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "geralfs messenger",
    "collectorNumber": undefined,
    "count": "4",
    "expansionCode": undefined,
    "type": "maindeck",
  },
  Object {
    "cardName": "damping sphere",
    "collectorNumber": undefined,
    "count": "3",
    "expansionCode": undefined,
    "type": "sideboard",
  },
]
`;
